exporters/prometheus/src/exporter_utils.cc

```diff
@@ -20,6 +20,69 @@ namespace exporter
 {
 namespace metrics
 {
+namespace
+{
+template <typename T>
+inline std::string Sanitize(std::string name, const T &valid)
+{
+  constexpr const auto replacement     = '_';
+  constexpr const auto replacement_dup = '=';
+
+  bool has_dup = false;
+  for (int i = 0; i < (int)name.size(); ++i)
+  {
+    if (valid(i, name[i]))
+    {
+      continue;
+    }
+    if (i > 0 && (name[i - 1] == replacement || name[i - 1] == replacement_dup))
+    {
+      has_dup = true;
+      name[i] = replacement_dup;
+    }
+    else
+    {
+      name[i] = replacement;
+    }
+  }
+  if (has_dup)
+  {
+    auto end = std::remove(name.begin(), name.end(), replacement_dup);
+    return std::string{name.begin(), end};
+  }
+  return name;
+}
+
+/**
+ * Sanitize the given metric label key according to Prometheus rule.
+ * Prometheus metric label keys are required to match the following regex:
+ * [a-zA-Z_]([a-zA-Z0-9_])*
+ */
+std::string SanitizeLabel(std::string label_key)
+{
+  return Sanitize(label_key, [](int i, char c) {
+    return (c >= 'a' && c <= 'z') ||  //
+           (c >= 'A' && c <= 'Z') ||  //
+           (c >= '0' && c <= '9' && i > 0);
```

Comment:
nit - the comment mentions underscore `_` in regex match, while it is missing in above conditional check.

Reply:
Thanks, that's a great catch. I didn't look very closely because I was adapting the existing implementation, which had the same bug.

Reply:
On second thought - the code is okay but for a subtle reason. It classifies `_` as invalid, then replaces it with `_` as part of sanitation.

I've made the change anyway for the sake of clarity.

Reply:
I felt the change is required otherwise valid string like `_ _ a` will be converted to `_ = a` and eventually to ` _ a` which is not expected :)

Reply:
But that *is* the desired behavior (collapse multiple consecutive underscores to a single underscore). @lalitb maybe I'm misunderstanding your comment.

Reply:
Yes, sorry for the confusion. I didn't read that part of the specs. Should be good.